### ISSUE TITLE
How to build the PyTorch Backend on jetson?

### ISSUE THREAD

**Issue by JulyLi2019** [triton-inference-server/server#3203]
At present, I have run the official code of Tensorflow Backend example on jetson, and then how to build the PyTorch Backend on jetson?What should i do next?

**Comment by JulyLi2019** [comment]
It is worth noting that the installation of tritonserver on jetson does not use docker.

**Comment by tanmayv25** [comment]
The Triton Jetson build is not docker based. We don't officially support pytorch backend on jetson but it may be possible by building it from scratch.
@CoderHam  Do you know if they can build pytorch backend on jetson?

**Comment by CoderHam** [comment]
@JulyLi2019 you should be able to build the pytorch backend. I do know of some users being able to successfully build the pytorch backend on Jetson. 
You will have to disable MKL in the cmake files and follow the build instructions here: https://github.com/triton-inference-server/pytorch_backend#build-the-pytorch-backend
In specific you will follow the non-docker build for pytorch backend.

You can follow the example [here](https://github.com/dusty-nv/jetson-containers/blob/triton/packages/tritonserver/pytorch_backend/CMakeLists.txt)
@dusty-nv can you share your insights on the same?

**Comment by JulyLi2019** [comment]
> @JulyLi2019您应该能够构建 pytorch 后端。我确实知道一些用户能够在 Jetson 上成功构建 pytorch 后端。
> 您必须在 cmake 文件中禁用 MKL 并按照此处的构建说明进行操作：[https](https://github.com/triton-inference-server/pytorch_backend#build-the-pytorch-backend) : [//github.com/triton-inference-server/pytorch_backend#build-the-pytorch-backend](https://github.com/triton-inference-server/pytorch_backend#build-the-pytorch-backend)
> 具体而言，您将遵循非 docker 构建用于 pytorch 后端。
> 
> 你可以按照[这里](https://github.com/dusty-nv/jetson-containers/blob/triton/packages/tritonserver/pytorch_backend/CMakeLists.txt)的例子
> @dusty-nv你能分享你的见解吗？

Now, I used the CMakeLists.txt file you gave to run`cmake -DCMAKE_INSTALL_PREFIX:PATH=`pwd`/install -DTRITON_PYTORCH_INCLUDE_PATHS="<PATH_PREFIX>/torch;<PATH_PREFIX>/torch/torch/csrc/api/include;<PATH_PREFIX>/torchvision" -DTRITON_PYTORCH_LIB_PATHS="<LIB_PATH_PREFIX>" ..`,system hint:`  By not providing "FindPython3.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Python3", but
  CMake did not find one.

  Could not find a package configuration file provided by "Python3" with any
  of the following names:

    Python3Config.cmake
    python3-config.cmake

  Add the installation prefix of "Python3" to CMAKE_PREFIX_PATH or set
  "Python3_DIR" to a directory containing one of the above files.  If
  "Python3" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
`,how should I do next? By the way,since jetson cmake version is 3.10.2,I changed the cmake version in CMakeLists.txt to 3.10.2.

**Comment by CoderHam** [comment]
You should upgrade your cmake version to 3.18 instead of downgrading the min version for the backend. We chose to upgrade the min version for a reason.

You will also need to install the necessary dependencies for the pytorch backend including torchvision and python3.

**Comment by JulyLi2019** [comment]
Now,I have upgrade my cmake version to 3.21.0,and run the `cmake -DCMAKE_INSTALL_PREFIX:PATH=`pwd`/install -DTRITON_PYTORCH_INCLUDE_PATHS="<PATH_PREFIX>/torch;<PATH_PREFIX>/torch/torch/csrc/api/include;<PATH_PREFIX>/torchvision" -DTRITON_PYTORCH_LIB_PATHS="<LIB_PATH_PREFIX>" ..`
,but it gives the following error
`-- RapidJSON found. Headers: /usr/include
-- RapidJSON found. Headers: /usr/include
-- Using CUDA 10.2
CMake Error: File /home/nvidia/Desktop/LH/src/libtriton_pytorch.ldscript does not exist.
CMake Error at CMakeLists.txt:88 (configure_file):
  configure_file Problem configuring file


CMake Error: File /home/nvidia/Desktop/LH/cmake/TritonPyTorchBackendConfig.cmake.in does not exist.
CMake Error at /opt/cmake-3.21.0/share/cmake-3.21/Modules/CMakePackageConfigHelpers.cmake:342 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  CMakeLists.txt:200 (configure_package_config_file)


-- Configuring incomplete, errors occurred!
See also "/home/nvidia/Desktop/LH/build/CMakeFiles/CMakeOutput.log".
See also "/home/nvidia/Desktop/LH/build/CMakeFiles/CMakeError.log".
`
how should i do  next?@CoderHam

**Comment by CoderHam** [comment]
@JulyLi2019 looks like you are not setting your build directory/paths correctly. Are you running this cmake command inside pytorch_backend/build?

**Comment by JulyLi2019** [comment]
> @JulyLi2019 looks like you are not setting your build directory/paths correctly. Are you running this cmake command inside pytorch_backend/build?

@CoderHam Now,I have ran this cmake command  `cmake -DCMAKE_INSTALL_PREFIX:PATH=`pwd`/install -DTRITON_PYTORCH_INCLUDE_PATHS="<PATH_PREFIX>/torch;<PATH_PREFIX>/torch/torch/csrc/api/include;<PATH_PREFIX>/torchvision" -DTRITON_PYTORCH_LIB_PATHS="<LIB_PATH_PREFIX>" ..`inside pytorch_backend/build,it ran successfully.
But,when I run `make install`,system hit
`fatal error: torch/script.h: No such file or directory
 #include <torch/script.h>  // One-stop header for TorchScript
          ^~~~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/triton-pytorch-backend.dir/build.make:75: recipe for target 'CMakeFiles/triton-pytorch-backend.dir/src/libtorch.cc.o' failed
make[2]: *** [CMakeFiles/triton-pytorch-backend.dir/src/libtorch.cc.o] Error 1
CMakeFiles/Makefile2:144: recipe for target 'CMakeFiles/triton-pytorch-backend.dir/all' failed
make[1]: *** [CMakeFiles/triton-pytorch-backend.dir/all] Error 2
Makefile:135: recipe for target 'all' failed
make: *** [all] Error 2
`
How should I solve this problem?

**Comment by JulyLi2019** [comment]
> @JulyLi2019 looks like you are not setting your build directory/paths correctly. Are you running this cmake command inside pytorch_backend/build?

@CoderHam The main problem should be: torch/script.h: No such file or directory.

**Comment by CoderHam** [comment]
You need to link against the pytorch library correctly. Check your paths to ensure "<PATH_PREFIX>/torch is not empty

**Comment by CoderHam** [comment]
Closing due to inactivity. Re-open if needed.

**Comment by julienripoche** [comment]
Hi, is there any plan to support the pytorch backend on jetson in a near future ? Or should we still try to build it by hand ?

**Comment by CoderHam** [comment]
@julienripoche you should be able to build it by hand. We do not have a final date for official pytorch on jetson support but the steps above along with the comment in the Cmakelist.txt should be sufficient to build the jetson pytorch backend.

**Comment by julienripoche** [comment]
@CoderHam thanks for the reply !
I managed to build it for tritonserver 2.13.0 (jetpack 4.6).
Quick question though, I was used to use `platform: "pytorch_libtorch"` in the config.pbtxt on non-arm architecture.
But on the jetson I got this error when launching the triton server:
`E1106 12:05:05.851903 21334 model_repository_manager.cc:1919] Poll failed for model directory '***': unexpected platform type pytorch_libtorch for ***`
When I switch to `backend: "pytorch"` in the config.pbtxt it works fine.
Any idea on why I should backend instead of platform ?
